In [28]:
#import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandasql import sqldf



In [ ]:
#load dataset
df=pd.read_excel("c:\\Users\\User\\Downloads\\customer_data.xlsx")
df

customer_id  gender  age payment_method
0         C241288  Female   28    Credit Card
1         C111565    Male   21     Debit Card
2         C266599    Male   20           Cash
3         C988172  Female   66    Credit Card
4         C189076  Female   53           Cash
...           ...     ...  ...            ...
99452     C441542  Female   45    Credit Card
99453     C569580    Male   27           Cash
99454     C103292    Male   63     Debit Card
99455     C800631    Male   56           Cash
99456     C273973  Female   36    Credit Card

[99457 rows x 4 columns]

In [ ]:
#data to dataframe
DF2=pd.DataFrame(df)

In [31]:
DF2

customer_id  gender  age payment_method
0         C241288  Female   28    Credit Card
1         C111565    Male   21     Debit Card
2         C266599    Male   20           Cash
3         C988172  Female   66    Credit Card
4         C189076  Female   53           Cash
...           ...     ...  ...            ...
99452     C441542  Female   45    Credit Card
99453     C569580    Male   27           Cash
99454     C103292    Male   63     Debit Card
99455     C800631    Male   56           Cash
99456     C273973  Female   36    Credit Card

[99457 rows x 4 columns]

In [ ]:
#head
Q="SELECT * FROM DF2 LIMIT 5"
RESULT=sqldf(Q,locals())
RESULT

customer_id  gender  age payment_method
0     C241288  Female   28    Credit Card
1     C111565    Male   21     Debit Card
2     C266599    Male   20           Cash
3     C988172  Female   66    Credit Card
4     C189076  Female   53           Cash

In [ ]:
#tail
Q="SELECT * FROM DF2 ORDER BY ROWID DESC LIMIT 5"
RESULT=sqldf(Q,locals())
RESULT

customer_id  gender  age payment_method
0     C273973  Female   36    Credit Card
1     C800631    Male   56           Cash
2     C103292    Male   63     Debit Card
3     C569580    Male   27           Cash
4     C441542  Female   45    Credit Card

In [ ]:
#data cleaning
Q="SELECT * FROM DF2 WHERE customer_id IS NULL"
RESULT=sqldf(Q,locals())
RESULT

Empty DataFrame
Columns: [customer_id, gender, age, payment_method]
Index: []

In [35]:
Q="SELECT * FROM DF2 WHERE gender IS NULL"
RESULT=sqldf(Q,locals())
RESULT

Empty DataFrame
Columns: [customer_id, gender, age, payment_method]
Index: []

In [36]:
Q="SELECT *FROM DF2 WHERE age IS NULL;"
RESULT=sqldf(Q,locals())
RESULT


Empty DataFrame
Columns: [customer_id, gender, age, payment_method]
Index: []

In [37]:
Q="SELECT * FROM DF2 WHERE payment_method IS NULL;"
RESULT=sqldf(Q,locals())
RESULT

Empty DataFrame
Columns: [customer_id, gender, age, payment_method]
Index: []

In [ ]:
#Retrieve Customers Based on Specific Age
Q="SELECT * FROM DF2 WHERE age >68;"
RESULT=sqldf(Q,locals())
RESULT

customer_id  gender  age payment_method
0        C159642    Male   69    Credit Card
1        C135790  Female   69           Cash
2        C171154  Female   69           Cash
3        C848969    Male   69           Cash
4        C336232    Male   69           Cash
...          ...     ...  ...            ...
1892     C203641  Female   69           Cash
1893     C685875    Male   69           Cash
1894     C955663  Female   69    Credit Card
1895     C274732  Female   69           Cash
1896     C321665    Male   69     Debit Card

[1897 rows x 4 columns]

In [ ]:
#Find Customers with a Specific Payment Method
Q="SELECT * FROM DF2 WHERE payment_method ='Credit Card';"
RESULT=sqldf(Q,locals())
RESULT

customer_id  gender  age payment_method
0         C241288  Female   28    Credit Card
1         C988172  Female   66    Credit Card
2         C657758  Female   28    Credit Card
3         C176086  Female   32    Credit Card
4         C159642    Male   69    Credit Card
...           ...     ...  ...            ...
34926     C142643    Male   28    Credit Card
34927     C363191  Female   47    Credit Card
34928     C231387  Female   65    Credit Card
34929     C441542  Female   45    Credit Card
34930     C273973  Female   36    Credit Card

[34931 rows x 4 columns]

In [ ]:
#Count the Number of Customers in Each Payment Method Category
Q="SELECT payment_method, COUNT(*) AS customer_count FROM DF2 GROUP BY payment_method;"
RESULT=sqldf(Q,locals())
RESULT

payment_method  customer_count
0           Cash           44447
1    Credit Card           34931
2     Debit Card           20079

In [ ]:
#Find Customers in a Specific Age Range
Q="SELECT * FROM DF2 WHERE age BETWEEN 65 AND 69;"
RESULT=sqldf(Q,locals())
RESULT

customer_id  gender  age payment_method
0        C988172  Female   66    Credit Card
1        C159642    Male   69    Credit Card
2        C225330  Female   67     Debit Card
3        C555402  Female   67    Credit Card
4        C300786    Male   65     Debit Card
...          ...     ...  ...            ...
9405     C156575  Female   67           Cash
9406     C241121  Female   67           Cash
9407     C221548  Female   66           Cash
9408     C231387  Female   65    Credit Card
9409     C953724    Male   65           Cash

[9410 rows x 4 columns]

In [ ]:
#Count the Number of Male and Female Customers
Q="SELECT gender, COUNT(*) AS customer_count FROM DF2 GROUP BY gender;"
RESULT=sqldf(Q,locals())
RESULT

gender  customer_count
0  Female           59482
1    Male           39975

In [ ]:
#Find the Oldest Customer
Q="SELECT * FROM DF2 ORDER BY age DESC LIMIT 1;"
RESULT=sqldf(Q,locals())
RESULT

customer_id gender  age payment_method
0     C159642   Male   69    Credit Card

In [ ]:
#Find the Youngest Customer
Q = "SELECT * FROM DF2 ORDER BY age ASC LIMIT 1;"
RESULT=sqldf(Q,locals())
RESULT

customer_id gender  age payment_method
0     C126436   Male   18     Debit Card

In [ ]:
#Find the Total Number of Customers (Count All Rows)
Q="SELECT COUNT(*) AS total_customer FROM DF2;"
RESULT=sqldf(Q,locals())
RESULT

total_customer
0           99457

In [ ]:
#Find customer_id and payment_method
Q="SELECT customer_id, payment_method FROM DF2;"
RESULT=sqldf(Q,locals())
RESULT

customer_id payment_method
0         C241288    Credit Card
1         C111565     Debit Card
2         C266599           Cash
3         C988172    Credit Card
4         C189076           Cash
...           ...            ...
99452     C441542    Credit Card
99453     C569580           Cash
99454     C103292     Debit Card
99455     C800631           Cash
99456     C273973    Credit Card

[99457 rows x 2 columns]

In [ ]:
#Find Customers Who Have Used Multiple Payment Methods
Q="SELECT customer_id FROM DF2 GROUP BY customer_id HAVING COUNT(DISTINCT payment_method) >1;"
RESULT=sqldf(Q,locals())
RESULT

Empty DataFrame
Columns: [customer_id]
Index: []

In [ ]:
#List All Unique Payment Methods Used
Q="SELECT DISTINCT payment_method FROM DF2;"
RESULT=sqldf(Q,locals())
RESULT

payment_method
0    Credit Card
1     Debit Card
2           Cash

In [ ]:
#Find Customers Who Have Used Only One Payment Method
Q="SELECT customer_id FROM DF2 GROUP BY customer_id HAVING COUNT(DISTINCT payment_method) =1;"
RESULT=sqldf(Q,locals())
RESULT

customer_id
0         C100004
1         C100005
2         C100006
3         C100012
4         C100019
...           ...
99452     C999886
99453     C999910
99454     C999974
99455     C999976
99456     C999995

[99457 rows x 1 columns]

In [ ]:
#Find the Average Age of Customers
Q="SELECT AVG(age) AS average_age FROM DF2;"
RESULT=sqldf(Q,locals())
RESULT

average_age
0    43.438954

In [ ]:
#List Customers Who Are Below 19 and Use Credit Card
Q="SELECT * FROM DF2 WHERE age<19 AND payment_method='Credit Card';"
RESULT=sqldf(Q,locals())
RESULT

customer_id  gender  age payment_method
0       C166635    Male   18    Credit Card
1       C246719  Female   18    Credit Card
2       C800785  Female   18    Credit Card
3       C382952    Male   18    Credit Card
4       C104728  Female   18    Credit Card
..          ...     ...  ...            ...
645     C247415  Female   18    Credit Card
646     C244904    Male   18    Credit Card
647     C231280    Male   18    Credit Card
648     C259264  Female   18    Credit Card
649     C250277  Female   18    Credit Card

[650 rows x 4 columns]

In [ ]:
#Get the Customer with the Most Common Payment Method
Q="SELECT payment_method,COUNT(*) AS count FROM DF2 WHERE age BETWEEN 30 AND 40 GROUP BY payment_method ORDER BY count DESC LIMIT 1;"
RESULT=sqldf(Q,locals())
RESULT

payment_method  count
0           Cash   9461

In [ ]:
#Rank Payment Methods Based on Usage
Q="SELECT payment_method, COUNT(*) AS transactionS,RANK() OVER(ORDER BY COUNT(*)DESC) AS RANK FROM DF2 GROUP BY payment_method;"
RESULT=sqldf(Q,locals())
RESULT

payment_method  transactionS  RANK
0           Cash         44447     1
1    Credit Card         34931     2
2     Debit Card         20079     3

In [ ]:
#Find the Percentage of Male and Female Customers
Q="SELECT gender ,COUNT(*) * 100.0/ SUM(COUNT(*)) OVER() AS percentage FROM DF2 GROUP BY GENDER;"
RESULT=sqldf(Q,locals())
RESULT

gender  percentage
0  Female   59.806751
1    Male   40.193249

CONCLUSION

1.Most customers are female (60%), and the average age is 43 years.

2.The most used payment method is Cash (45%), followed by Credit Card (35%) and Debit Card (20%).